examples/rp/src/bin/pio_uart.rs

```diff
@@ -0,0 +1,431 @@
+    use fixed_macro::types::U56F8;
+
+    use crate::uart::PioUartError;
+
+    pub struct PioUartTx<'a> {
+        sm_tx: StateMachine<'a, PIO0, 0>,
+    }
+
+    impl<'a> PioUartTx<'a> {
+        pub fn new(
+            common: &mut Common<'a, PIO0>,
+            mut sm_tx: StateMachine<'a, PIO0, 0>,
+            tx_pin: impl PioPin,
+            baud: u64,
+            origin: Option<u8>,
+        ) -> (Self, u8) {
+            let mut prg = pio_proc::pio_asm!(
+                r#"
+                .side_set 1 opt
+
+                ; An 8n1 UART transmit program.
+                ; OUT pin 0 and side-set pin 0 are both mapped to UART TX pin.
+
+                    pull       side 1 [7]  ; Assert stop bit, or stall with line in idle state
+                    set x, 7   side 0 [7]  ; Preload bit counter, assert start bit for 8 clocks
+                bitloop:                   ; This loop will run 8 times (8n1 UART)
+                    out pins, 1            ; Shift 1 bit from OSR to the first OUT pin
+                    jmp x-- bitloop   [6]  ; Each loop iteration is 8 cycles.
+            "#
+            );
+            prg.program.origin = origin;
+            let tx_pin = common.make_pio_pin(tx_pin);
+            sm_tx.set_pins(Level::High, &[&tx_pin]);
+            sm_tx.set_pin_dirs(Direction::Out, &[&tx_pin]);
+
+            let relocated = RelocatedProgram::new(&prg.program);
+
+            let mut cfg = Config::default();
+
+            cfg.set_out_pins(&[&tx_pin]);
+            cfg.use_program(&common.load_program(&relocated), &[&tx_pin]);
+            cfg.shift_out.auto_fill = false;
+            cfg.shift_out.direction = ShiftDirection::Right;
+            cfg.fifo_join = FifoJoin::TxOnly;
+            cfg.clock_divider = (U56F8!(125_000_000) / (8 * baud)).to_fixed();
+            sm_tx.set_config(&cfg);
+            sm_tx.set_enable(true);
+
+            // The 4 state machines of the PIO each have their own program counter that starts taking
+            // instructions at an offset (origin) of the 32 instruction "space" the PIO device has.
+            // It is up to the programmer to sort out where to place these instructions.
+            // From the pio_asm! macro you get a ProgramWithDefines which has a field .program.origin
+            // which takes an Option<u8>.
+            //
+            // When you load more than one RelocatedProgram into the PIO,
+            // you load your first program at origin = 0.
+            // The RelocatedProgram has .code().count() which returns a usize,
+            // for which you can then use as your next program's origin.
+            let offset = relocated.code().count() as u8 + origin.unwrap_or_default();
+            (Self { sm_tx }, offset)
+        }
+
+        pub async fn write_u8(&mut self, data: u8) {
+            self.sm_tx.tx().wait_push(data as u32).await;
+        }
+    }
+
+    impl Io for PioUartTx<'_> {
+        type Error = PioUartError;
```

Comment:
if reading/writing never fails, the convention is to do `type Error = core::convert::Infallible`. It's an uninhabited enum (0 variants) which means it's impossible to create a value of it.

Reply:
Fixed